libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+            // build from another cobuild in this case.
+            const cobuildCompletedState: ICobuildCompletedState | undefined =
+              await cobuildLock.getCompletedStateAsync();
+            if (cobuildCompletedState) {
+              const { status, cacheId } = cobuildCompletedState;
+
+              const restoreFromCacheSuccess: boolean | undefined =
+                await cobuildLock.projectBuildCache.tryRestoreFromCacheAsync(buildCacheTerminal, cacheId);
+
+              if (restoreFromCacheSuccess) {
+                // Restore the original state of the operation without cache
+                await context._operationMetadataManager?.tryRestoreAsync({
+                  terminal,
+                  logPath,
+                  errorLogPath
+                });
+                if (cobuildCompletedState) {
+                  return cobuildCompletedState.status;
+                }
+                return status;
+              }
+            }
+          } else if (buildCacheContext.isCacheReadAllowed) {
+            buildCacheReadAttempted = !!projectBuildCache;
+            const restoreFromCacheSuccess: boolean | undefined =
+              await projectBuildCache?.tryRestoreFromCacheAsync(buildCacheTerminal);
+
+            if (restoreFromCacheSuccess) {
+              // Restore the original state of the operation without cache
+              await context._operationMetadataManager?.tryRestoreAsync({
+                terminal,
+                logPath,
+                errorLogPath
+              });
+              return OperationStatus.FromCache;
+            }
+          }
+          if (buildCacheContext.isSkipAllowed && !buildCacheReadAttempted) {
+            const isPackageUnchanged: boolean = !!(
+              lastProjectDeps &&
+              projectDeps &&
+              projectDeps.arguments === lastProjectDeps.arguments &&
+              _areShallowEqual(projectDeps.files, lastProjectDeps.files)
+            );
+
+            if (isPackageUnchanged) {
+              return OperationStatus.Skipped;
+            }
+          }
+
+          if (buildCacheContext.isCacheWriteAllowed && cobuildLock) {
+            const acquireSuccess: boolean = await cobuildLock.tryAcquireLockAsync();
+            if (acquireSuccess) {
+              // The operation may be used to marked remote executing, now change it to executing
+              context.status = OperationStatus.Executing;
+              runner.periodicCallback.addCallback(async () => {
+                await cobuildLock?.renewLockAsync();
+              });
+            } else {
+              // failed to acquire the lock, mark current operation to remote executing
+              context.stopwatch.reset();
+              return OperationStatus.RemoteExecuting;
+            }
+          }
+        })();
+        if (earlyReturnStatus) {
+          beforeExecuteContext.earlyReturnStatus = earlyReturnStatus;
+        }
+        return beforeExecuteContext;
```

Comment:
Rather than being a waterfall returning the context object (which is mutable and therefore doesn't need to be returned), this should be a bail hook and return `earlyReturnStatus`.